sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateViewAsSelect.scala

```diff
@@ -73,15 +68,18 @@ private[hive] case class CreateViewAsSelect(
   }
 
   private def prepareTable(sqlContext: SQLContext): CatalogTable = {
-    val expandedText = if (sqlContext.conf.canonicalView) {
-      try rebuildViewQueryString(sqlContext) catch {
-        case NonFatal(e) => wrapViewTextWithSelect
-      }
+    val logicalPlan = if (tableDesc.schema.isEmpty) {
+      child
     } else {
-      wrapViewTextWithSelect
+      val projectList = childSchema.zip(tableDesc.schema).map {
+        case (attr, col) => Alias(attr, col.name)()
+      }
+      sqlContext.executePlan(Project(projectList, child)).analyzed
     }
 
-    val viewSchema = {
+    val viewText = new SQLBuilder(logicalPlan, sqlContext).toSQL
```

Comment:
How about catching exceptions thrown from `toSQL` and suggest the user to turn off native view and retry?